drivers/clock_control/nrf_power_clock.c

```diff
@@ -499,38 +469,103 @@ void z_nrf_clock_control_lf_on(enum nrf_lfclk_start_mode start_mode)
- * particular name.  The platform architecture itself doesn't care,
- * but there is a test (tests/kernel/arm_irq_vector_table) that needs
- * to find it to it can set it in a custom vector table.  Should
- * probably better abstract that at some point (e.g. query and reset
- * it by pointer at runtime, maybe?) so we don't have this leaky
- * symbol.
- */
-void nrf_power_clock_isr(void *arg);
+static void clock_event_handler(nrfx_clock_evt_type_t event)
+{
+	struct device *dev = DEVICE_GET(clock_nrf);
+	switch (event) {
+	case NRFX_CLOCK_EVT_HFCLK_STARTED:
+	{
+		struct nrf_clock_control_sub_data *data =
+				get_sub_data(dev, CLOCK_CONTROL_NRF_TYPE_HFCLK);
+
+		/* Check needed due to anomaly 201:
+		 * HFCLKSTARTED may be generated twice.
+		 */
+		if (GET_STATUS(data->flags) == CLOCK_CONTROL_STATUS_STARTING) {
+			clkstarted_handle(dev, CLOCK_CONTROL_NRF_TYPE_HFCLK);
+		}
+
+		break;
+	}
+	case NRFX_CLOCK_EVT_LFCLK_STARTED:
+		if (IS_ENABLED(
+			CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION)) {
+			z_nrf_clock_calibration_lfclk_started();
+		}
+		clkstarted_handle(dev, CLOCK_CONTROL_NRF_TYPE_LFCLK);
+		break;
+	case NRFX_CLOCK_EVT_CAL_DONE:
+		if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION)) {
+			z_nrf_clock_calibration_done_handler();
+		}
+		break;
+	default:
+		__ASSERT_NO_MSG(0);
+		break;
+	}
+}
+
+#if defined(CONFIG_USB_NRFX)
+static void usb_event_handler(nrfx_power_usb_evt_t event)
+{
+	extern void usb_dc_nrfx_power_event_callback(nrfx_power_usb_evt_t event);
+
+	switch (event) {
+	case NRFX_POWER_USB_EVT_DETECTED:
+		usb_dc_nrfx_power_event_callback(NRFX_POWER_USB_EVT_DETECTED);
+		break;
+	case NRFX_POWER_USB_EVT_REMOVED:
+		usb_dc_nrfx_power_event_callback(NRFX_POWER_USB_EVT_REMOVED);
+		break;
+	case NRFX_POWER_USB_EVT_READY:
+		usb_dc_nrfx_power_event_callback(NRFX_POWER_USB_EVT_READY);
+		break;
+	default:
+		__ASSERT_NO_MSG(0);
+		break;
+	}
+}
+#endif
 
 static int clk_init(struct device *dev)
 {
+	nrfx_err_t nrfx_err;
 	int err;
 	static const struct onoff_transitions transitions = {
 		.start = onoff_start,
 		.stop = onoff_stop
 	};
 
 	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
-		    nrf_power_clock_isr, 0, 0);
-
+		    nrfx_isr, nrfx_power_clock_irq_handler, 0);
 	irq_enable(DT_INST_IRQN(0));
 
-	nrf_clock_lf_src_set(NRF_CLOCK, CLOCK_CONTROL_NRF_K32SRC);
+#if (defined(CONFIG_USB_NRFX) && defined(CONFIG_SOC_SERIES_NRF53X))
+	IRQ_CONNECT(USBREGULATOR_IRQn, DT_INST_IRQ(0, priority),
+		    nrfx_usbreg_irq_handler, 0, 0);
+	irq_enable(USBREGULATOR_IRQn);
+#endif
+
+	nrfx_err = nrfx_clock_init(clock_event_handler);
+	if (nrfx_err != NRFX_SUCCESS) {
+		return -EIO;
+	}
 
 	if (IS_ENABLED(CONFIG_CLOCK_CONTROL_NRF_K32SRC_RC_CALIBRATION)) {
 		struct nrf_clock_control_data *data = dev->driver_data;
 
 		z_nrf_clock_calibration_init(data->mgr);
 	}
 
-	clock_irqs_enable();
+	nrfx_clock_enable();
+
+#if defined(CONFIG_USB_NRFX)
+	static const nrfx_power_usbevt_config_t config = {
```

Comment:
This structure does not need to be `static`.

Reply:
it can land in flash only if it is static. Compiling samples/subsys/cdc_acm with and without static show FLASH +12 after removing `static`.

Reply:
That's strange. `static` should not matter in this regard. As long as it is `const` it can be placed in flash, if that's suitable.

Reply:
noticed it already in the past, seems that destination rule (static -> fixed address, no static ->stack) is evaluated before const keyword.